### PR TITLE
Implement Full CRUD Operations with Redis

### PR DESCRIPTION
**Pull request by TC-ibautista**
### Summary
This pull request introduces full CRUD operations for items using Redis as the data store. It includes the following changes:
- **Create Item**: Implemented POST method to create a new item.
- **Read All Items**: Implemented GET method to retrieve all items.
- **Read Single Item**: Implemented GET method to retrieve a specific item by name.
- **Update Item**: Implemented PUT method to update an existing item.
- **Delete Item**: Implemented DELETE method to remove an item.
- **Partial Update**: Implemented PATCH method to update specific attributes of an item, such as the price.

### Testing
- Tested all CRUD operations using appropriate HTTP methods.

### Notes
- Ensure that the Redis server is running and accessible.
- Update environment variables `REDIS_HOST` and `REDIS_PORT` as needed.

### Feedback
- I would greatly appreciate any feedback you can provide to help improve my code. 
- Additionally, if anything seems unclear or doesn't make sense, please don't hesitate to ask for clarification.

### CODE REVIEW

**Comment by TC-jBautista** [comment]
Test

File: erlang/api_cowboy/.gitignore
Comment (TC-jCruz) [inline erlang/api_cowboy/.gitignore:1]: If you're using Visual Studio Code I recommend you use Erlang LS extension, and for that I'll generate an `erlang_ls.config` that you must add to the `.gitignore` fiel too, just a FYI, you don't have to use it if you don't want but It helps a lot

File: erlang/api_cowboy/README.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+api_cowboy
+=====
+
+An OTP application
+
+Build
+-----
+
+    $ rebar3 compile
+    $ rebar3 shell
```
Comment (TC-jCruz) [inline erlang/api_cowboy/README.md]: Would recommend to be more explicit on what does each command does.

File: erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    {ok, Req, State}.
+
+handle_method_not_allowed(Req0, State) ->
+    Req = cowboy_req:reply(405,
+        #{<<"content-type">> => <<"text/plain">>},
+        <<"Method Not Allowed">>,
+        Req0),
+    {ok, Req, State}.
+
+connect_to_redis() ->
```
Comment (TC-Haguilar) [inline erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl]: I recommend that you make a new redis module with all the operations (add, edit, delete, etc), in addition to the fact that each request is running a new redis process

File: erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            handle_post_items(Req0, State);
+        {<<"PUT">>, <<"/api/v1/items/", ItemName/binary>>} ->
+            handle_put_item(Req0, State, ItemName);
+        {<<"PATCH">>, <<"/api/v1/items/", ItemName/binary>>} ->
+            handle_patch_item_price(Req0, State, ItemName);
+        {<<"DELETE">>, <<"/api/v1/items/", ItemName/binary>>} ->
+            handle_delete_item(Req0, State, ItemName);
+        _ ->
+            handle_method_not_allowed(Req0, State)
+    end.
```
Comment (TC-jCruz) [inline erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl]: As I told you yesterday, this should not be defined this way.
Cowboy will route a request to you defined callback if you define it correctly.
I would suggest you to give it another look to the request handling diagram https://ninenines.eu/docs/en/cowboy/2.12/guide/rest_flowcharts/

We can schedule a meeting if you want us to check it further

File: erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+handle_root(Req0, State) ->
+    Req = cowboy_req:reply(200,
+        #{<<"content-type">> => <<"text/plain">>},
+        <<"Welcome to the API Cowboy!">>,
+        Req0),
+    {ok, Req, State}.
+
+handle_get_items(Req0, State) ->
+    {ok, Client} = connect_to_redis(),
+    {ok, Keys} = eredis:q(Client, ["KEYS", "*"]),
```
Comment (TC-jCruz) [inline erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl]: It's preferable to define a specific module to handle the DB operations and handle all the reds related logic like the keys. It's not a good definition to have a handler or a controller handling logic.

File: erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        Req0),
+    {ok, Req, State}.
+
+handle_delete_item(Req0, State, ItemName) ->
+    {ok, Client} = connect_to_redis(),
+    {ok, DeletedCount} = eredis:q(Client, ["DEL", ItemName]),
+    Result = case DeletedCount of
+        <<"0">> -> <<"Item ", ItemName/binary, " not found">>;
+        <<"1">> -> <<"Item ", ItemName/binary, " deleted successfully">>;
+        _ -> <<"Unexpected result: ", DeletedCount/binary>>
```
Comment (TC-jCruz) [inline erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl]: I recommend this logic to be separated from the handler as the handler shall only handle the http request, not the logic of your domain itself

File: erlang/api_cowboy/.gitignore [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ebin
+log
+erl_crash.dump
+.rebar
+logs
+.idea
+*.iml
+rebar3.crashdump
+*~
+*.DS_Store
```
Comment (TC-eReyna) [inline erlang/api_cowboy/.gitignore]: You should enable the setting `Files: Insert Final Newline` in VSCode to get rid of those files without a new line at EOF.

File: erlang/api_cowboy/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,18 @@
+# Usar una imagen base de Erlang
```
Comment (TC-eReyna) [inline erlang/api_cowboy/Dockerfile:1]: A dockerfile should be useonly to deploy your application to production, in this case, you are using it to run your project locally and that's not the ideal. The only file that should run your project locally is your compose file.

I recommend the following:

1. Change the Dockerfile into a release build, it shold run `rebar3 release`, copy the generated app to a specific dir, change the base image to one without erlang, and run the project with the generated script.
2. In your docker-compose.yml file, run the project without the `build: .` param, you can check `docker-compose-local.yaml` on xmpp for reference or ask me directly if you have any question about this.

File: erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_app.erl [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+-behaviour(application).
+
+-export([start/2, stop/1]).
+
+start(_StartType, _StartArgs) ->
+    Dispatch = cowboy_router:compile([
+        {'_', [
+            {"/", api_cowboy_handler, []},
+            {"/api/v1/items", api_cowboy_handler, []},
+            {"/api/v1/items/:name", api_cowboy_handler, []}
```
Comment (TC-eReyna) [inline erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_app.erl]: You could save this list in a variable named Routes or something like that for better readability

```erlang
Routes = [
  {"/", api_cowboy_handler, []}
],
 Dispatch = cowboy_router:compile([{'_', Routes}]).
```

File: erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl [hunk 9]
```diff
@@ -0,0 +1,145 @@
+-module(api_cowboy_handler).
+-behaviour(cowboy_handler).
+
+-export([init/2]).
+
+init(Req0, State) ->
```
Comment (TC-eReyna) [inline erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl]: As Jose said, you should be using the rest_handler of cowboy, also one important thing for your current implementation, you are not allowing OPTIONS method and that's VERY important in rest calls, but don't worry, rest adapter will take care of that for you.

File: erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+handle_method_not_allowed(Req0, State) ->
+    Req = cowboy_req:reply(405,
+        #{<<"content-type">> => <<"text/plain">>},
+        <<"Method Not Allowed">>,
+        Req0),
+    {ok, Req, State}.
+
+connect_to_redis() ->
+    RedisHost = os:getenv("REDIS_HOST"),
```
Comment (TC-eReyna) [inline erlang/api_cowboy/apps/api_cowboy/src/api_cowboy_handler.erl]: In addition to what Hector said, I reccomend you to put your config under a config file so you can do `application:get_env()` and you can then point that config to the env variable

**Review by TC-eReyna — COMMENTED** [review]
Good work so far, keep it going

**Pull request merged**